homeassistant/components/sms/gateway.py

```diff
@@ -15,6 +18,120 @@ class Gateway:
     def __init__(self, worker, hass):
         """Initialize the sms gateway."""
         self._worker = worker
+        self._loop = asyncio.get_running_loop()
+        self._hass = hass
+
+    async def init_async(self):
+        """Initialize the sms gateway asyncronisly."""
+        try:
+            await self._worker.SetIncomingSMSAsync()
+        except gammu.ERR_NOTSUPPORTED:
+            _LOGGER.warning("Your phone does not support incoming SMS notifications!")
+        else:
+            await self._worker.SetIncomingCallbackAsync(self.sms_callback)
+
+    def sms_callback(self, state_machine, callback_type, callback_data):
```

Comment:
This is called inside the event loop right? In that case don't call `self._loop.call_soon_threadsafe` but instead just call `hass.bus.async_fire(…)` directly.

Reply:
> This is called inside the event loop right? In that case don't call `self._loop.call_soon_threadsafe` but instead just call `hass.bus.async_fire(…)` directly.

This is called from the gammu worker thread, not the hass loop.

Reply:
I see, okay in that case can you use `hass.add_job` to schedule `_notify_incoming_sms` and make sure `_notify_incoming_sms` is annoted with `@callback` and then use `async_fire` inside there.